.github/workflows/tests.yml

```diff
@@ -23,4 +23,6 @@ jobs:
         run: npm ci
 
       - name: Run tests
-        run: npm test
+        run: node_modules/.bin/percy exec -- npm test
+        env:
+          PERCY_TOKEN: ${{ secrets.PERCY_TOKEN }}
```

Comment:
This secret won't be available to PRs from forks, so we'll need to think about how this works for external contributors.

At the very least, we'll need to make sure this fails gracefully (the tests still run and pass) – based on the documentation you added I'm guessing it will still run the tests with '[percy] Percy is not running, disabling snapshots' but would be good to check this.

We'll also need to think about how this works if we aren't able to check for visual regressions in external PRs until after they've been merged.

Reply:
That's a good point. I've checked that behaviour, and you're right, the tests are run without Percy generating snapshots and you see this in the output:

```
[percy] Skipping visual tests
[percy] Error: Missing Percy token
[percy] Running "npm test"
.... [normal test output here]
```

I'll add some documentation for that, and have a think about what the process might look like for that..

Reply:
@36degrees I've been digging into this a bit more, and I think the fact that the checks can't run on a PR from a fork means that we can't make Percy a required check either (it'll never run for PRs from forks, and I can't find a way to manually trigger any kind of Percy build from our side).

I think that means our only option is to not mark Percy as a required check and just be disciplined in trying to review the builds when we review. Can you think of any other alternatives?

Reply:
It would be significantly more work, but if we wanted to we could explore using the `pull_request_target` or `workflow_run` events, which run in the context of the base of the pull request (likely `main`) and have access to secrets.

A workflow might do something like:

1) When a PR is opened, a workflow triggered by the `pull_request` event compile the JS and CSS from that branch and saves them as an artefact
2) Kick off another workflow using the `workflow_run` event, which runs in the context of the base branch and has access to secrets
3) That workflow downloads the built JS and CSS and runs the review app using the compiled JS and CSS, then runs the tests to take snapshots
4) Pass the context of the original PR to Percy to allow it to associate it with the PR, add the status checks etc.

We'd need to be very careful to ensure that we're not introducing potential vulnerabilities to our workflows – there's a good article on this here: https://securitylab.github.com/research/github-actions-preventing-pwn-requests/

My initial thought is we should probably try the simple thing first and explore this route in the future if we find it's causing friction?